pennylane/templates/layers/quantum_number_preserving_u2.py

```diff
@@ -0,0 +1,216 @@
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the QuantumNumberPreserving Gate Fabric template.
+"""
+# pylint: disable-msg=too-many-branches,too-many-arguments,protected-access
+import numpy as np
+import pennylane as qml
+from pennylane.operation import Operation, AnyWires
+
+
+class QuantumNumberPreservingU2(Operation):
+    r"""Implements a local, expressive, and quantum-number-preserving ansatz using VQE circuit fabrics
+    proposed by Anselmetti *et al.* in `arXiv:2104.05692 <https://arxiv.org/abs/2104.05692>`_.
+
+    This template prepares the :math:`2M` qubits trial state, where `M` is the number of spatial orbitals.
+    It uses 4-local-nearest-neighbor-tessellation of alternating even and off spatial-orbitals-pair 2-parameter,
+    4-qubit :math:`\hat{Q}(\varphi, \theta)` gates. Each :math:`\hat{Q}` gate consists of a 1-parameter, 4-qubit spatial rotation
+    gate :math:`\text{QuantumNumberPreserving}_{OR}(\varphi)`, and  a 1-parameter, 4-qubit diagonal pair exchange gate :math:`\text{DoubleExcitation}(\theta)`
+    gate. In the :math:`\hat{Q}` gate we also allow inclusion of an optional constant :math:`\hat{\Pi}` gate,
+    with choices :math:`\hat{\Pi} \in \{\hat{I}, \text{QuantumNumberPreserving}_{OR}(\pi)\}`.
+
+    The circuit implementing the gate fabric layer for `M = 4` is shown below:
+
+    |
+
+    .. figure:: ../../_static/templates/layers/quantum_number_preserving_u2.png
+        :align: center
+        :width: 25%
+        :target: javascript:void(0);
+
+    |
+
+    The 2-parameter, 4-qubit :math:`\hat{Q}` gate is decomposed as follows:
+
+    |
+
+    .. figure:: ../../_static/templates/layers/quantum_number_preserving_decompos.png
+        :align: center
+        :width: 100%
+        :target: javascript:void(0);
+
+    |
+
+    Args:
+        weights (tensor_like): Array of weights of shape ``(L, D, 2)``\.
+            ``L`` is the number of gate fabric layers and :math:`D=M-1`
+            is the number of :math:`\hat{Q}(\varphi, \theta)` gates per layer.
+        wires (Iterable): wires that the template acts on
+        init_state (tensor_like): iterable or shape ``(len(wires),)`` tensor representing the Hartree-Fock state
+            used to initialize the wires
+        pi_gate_include (boolean): If `True`, sets the optional constant :math:`\Pi` to :math:`\text{QuantumNumberPreserving}_{OR}(\pi)`.
+            Default value is :math:`\hat{I}`.
+
+    .. UsageDetails::
+
+        #. The number of wires :math:`N` has to be equal to the number of
+           spin orbitals included in the active space.
+
+        #. The number of trainable parameters scales linearly with the number of layers as
+           :math:`2D(N-1)`.
```

Comment:
The number of layers was named `L` before. I think this equation should be modified.

Reply:
Fixed.